File: 2024/16/16.py
#!/usr/bin/env python3
import heapq
import sys

def parse_input():
    maze = []
    start_position = None
    end_position = None
    
    for row_idx, line in enumerate(sys.stdin):
        line = line.strip()
        if not line:
            break
        maze.append(list(line))
        
        # Find start and end positions
        if 'S' in line:
            start_col = line.index('S')
            start_position = (row_idx, start_col)
        if 'E' in line:
            end_col = line.index('E')
            end_position = (row_idx, end_col)
    
    return maze, start_position, end_position

def move_forward(maze, row, col, direction):
    directions = {
        'N': (-1, 0),
        'E': (0, 1),
        'S': (1, 0),
        'W': (0, -1)
    }
    
    delta_row, delta_col = directions[direction]
    new_row = row + delta_row
    new_col = col + delta_col
    
    # Check bounds and avoid walls
    if 0 <= new_row < len(maze) and 0 <= new_col < len(maze[0]) and maze[new_row][new_col] != '#':
        return (new_row, new_col)
    
    return None

def rotate(direction, rotation):
    directions = ['N', 'E', 'S', 'W']
    current_index = directions.index(direction)
    
    if rotation == 'L':
        new_index = (current_index - 1) % 4
    elif rotation == 'R':
        new_index = (current_index + 1) % 4
    
    return directions[new_index]

def find_min_score_and_tiles(maze, start_position, end_position):
    # Priority queue (min-heap)
    pq = []
    heapq.heappush(pq, (0, 'E', start_position, [start_position]))  # Initial score, direction, position
    
    # Visited set to avoid revisiting the same state
    visited = {}
    
    # Dictionary to keep track of tiles on paths with a specific score
    paths_num = {}
    paths = {}
    min_score = float('inf')
    
    while pq:
        score, direction, position, trajectory = heapq.heappop(pq)
        if score > min_score:
            continue

        if (position, direction) in visited:
            if visited[(position, direction)] < score:
                continue
        
        visited[(position, direction)] = score
        
        # Check if we have reached the end
        if position == end_position:
            min_score = min(min_score, score)
            if min_score not in paths:
                paths[min_score] = set()
                paths_num[min_score] = 0
            paths[min_score].update(trajectory)
            paths_num[min_score] += 1
            continue
        
        # Try rotating left and right without moving
        for rotation in ['L', 'R']:
            new_direction = rotate(direction, rotation)
            heapq.heappush(pq, (score + 1000, new_direction, position, list(trajectory)))
        
        # Try moving forward
        new_position = move_forward(maze, *position, direction)
        if new_position:
            heapq.heappush(pq, (score + 1, direction, new_position, list(trajectory) + [new_position]))
    
    # Find unique tiles on best paths with the minimum score
    tiles_on_best_paths = paths[min_score] if min_score in paths else []
    
    return min_score, len(tiles_on_best_paths)

def main():
    maze, start_position, end_position = parse_input()
    min_score, unique_tile_count = find_min_score_and_tiles(maze, start_position, end_position)
    
    print("Minimum score to reach the end:", min_score)
    print("Number of unique tiles on best paths:", unique_tile_count)

if __name__ == "__main__":
    main()

File: 2024/16/16.rs
use std::collections::{HashMap, HashSet};
use std::io::{self, BufRead};

fn parse_input() -> (Vec<Vec<char>>, (usize, usize), (usize, usize)) {
    let stdin = io::stdin();
    let mut maze: Vec<Vec<char>> = vec![];
    let mut start_position = None;
    let mut end_position = None;

    for (row_idx, line) in stdin.lock().lines().enumerate() {
        let line = line.expect("Failed to read line");
        if line.is_empty() {
            break;
        }
        maze.push(line.chars().collect());

        // Find start and end positions
        if let Some(col_idx) = line.find('S') {
            start_position = Some((row_idx, col_idx));
        }
        if let Some(col_idx) = line.find('E') {
            end_position = Some((row_idx, col_idx));
        }
    }

    (maze, start_position.expect("Start position not found"), end_position.expect("End position not found"))
}

fn move_forward(maze: &Vec<Vec<char>>, row: usize, col: usize, direction: char) -> Option<(usize, usize)> {
    let directions = match direction {
        'N' => (-1, 0),
        'E' => (0, 1),
        'S' => (1, 0),
        'W' => (0, -1),
        _ => return None,
    };

    let new_row = row as isize + directions.0;
    let new_col = col as isize + directions.1;

    if new_row >= 0 && new_row < maze.len() as isize
        && new_col >= 0 && new_col < maze[0].len() as isize
        && maze[new_row as usize][new_col as usize] != '#'
    {
        Some((new_row as usize, new_col as usize))
    } else {
        None
    }
}

fn rotate(direction: char, rotation: &str) -> char {
    let directions = vec!['N', 'E', 'S', 'W'];
    let current_index = directions.iter().position(|&d| d == direction).unwrap();
    let new_index = match rotation {
        "L" => (current_index + 3) % 4,
        "R" => (current_index + 1) % 4,
        _ => panic!("Invalid rotation"),
    };
    directions[new_index]
}

fn find_min_score_and_tiles(maze: &Vec<Vec<char>>, start_position: (usize, usize), end_position: (usize, usize)) -> (i32, usize) {
    use std::collections::BinaryHeap;

    #[derive(PartialEq, Eq)]
    struct State {
        score: i32,
        direction: char,
        position: (usize, usize),
        trajectory: Vec<(usize, usize)>,
    }

    impl Ord for State {
        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
            self.score.cmp(&other.score).reverse()
        }
    }

    impl PartialOrd for State {
        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
            Some(self.cmp(other))
        }
    }

    let mut pq = BinaryHeap::new();
    pq.push(State {
        score: 0,
        direction: 'E',
        position: start_position,
        trajectory: vec![start_position],
    });

    let mut visited: HashMap<(usize, usize, char), i32> = HashMap::new();
    let mut paths: HashMap<i32, HashSet<(usize, usize)>> = HashMap::new();

    let mut min_score = i32::MAX;

    while let Some(State { score, direction, position, trajectory }) = pq.pop() {
        if score > min_score {
            continue;
        }

        let key = (position.0, position.1, direction);
        if visited.contains_key(&key) && *visited.get(&key).unwrap() <= score {
            continue;
        }
        visited.insert(key, score);

        // Debug print current state
        println!("Position: {:?}, Direction: {}, Score: {}", position, direction, score);

        // Check if we have reached the end
        if position == end_position {
            min_score = min_score.min(score);
            paths.entry(min_score).or_insert_with(HashSet::new).extend(trajectory.iter().cloned());
            println!("Reached end with score {}, trajectory length: {}", score, trajectory.len());
        } else {
            // Try rotating left and right without moving
            for rotation in ["L", "R"] {
                let new_direction = rotate(direction, rotation);
                pq.push(State {
                    score: score + 1000,
                    direction: new_direction,
                    position,
                    trajectory: trajectory.clone(),
                });
            }

            // Try moving forward
            if let Some(new_position) = move_forward(maze, position.0, position.1, direction) {
                pq.push(State {
                    score: score + 1,
                    direction,
                    position: new_position,
                    trajectory: trajectory.iter().cloned().chain(std::iter::once(new_position)).collect(),
                });
            }
        }
    }

    let unique_tiles_on_best_paths = paths.get(&min_score).unwrap_or(&HashSet::new()).len();
    println!("Unique tiles on best paths: {}", unique_tiles_on_best_paths);
    (min_score, unique_tiles_on_best_paths)
}

fn main() {
    let (maze, start_position, end_position) = parse_input();
    let (min_score, unique_tile_count) = find_min_score_and_tiles(&maze, start_position, end_position);
    println!("Minimum score to reach the end: {}", min_score);
    println!("Number of unique tiles on best paths: {}", unique_tile_count);
}